Compiler crashes on dependent type in an enum case

## minimized code

```Scala
trait A
  type B

enum Test
  case Test(a: A, b: a.B)
```


<details>
<summary>Stack trace</summary>

```scala
core.compile java.lang.AssertionError: assertion failed: orphan parameter reference: TermParamRef(a)
    dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
    dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:265)
    dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:140)
    dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:192)
    dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:140)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:542)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:291)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$37$$anonfun$1(TreePickler.scala:389)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$3(TreePickler.scala:389)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:390)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$36$$anonfun$1(TreePickler.scala:382)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$2(TreePickler.scala:382)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:383)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTreeUnlessEmpty(TreePickler.scala:294)
    dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$2(TreePickler.scala:308)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:310)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:504)
    dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:329)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:329)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$25(TreePickler.scala:530)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:531)
    dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$2(TreePickler.scala:305)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:310)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:506)
    dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:329)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:329)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$25(TreePickler.scala:530)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:531)
    dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$2(TreePickler.scala:305)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:310)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:506)
    dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:329)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:329)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$27(TreePickler.scala:540)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:44)
    dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:540)
    dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:695)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:695)
    dotty.tools.dotc.transform.Pickler.run$$anonfun$10$$anonfun$8(Pickler.scala:63)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.transform.Pickler.run$$anonfun$2(Pickler.scala:87)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.immutable.List.foreach(List.scala:305)
    dotty.tools.dotc.transform.Pickler.run(Pickler.scala:87)
    dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:315)
    scala.collection.immutable.List.map(List.scala:219)
    dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:316)
    dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:91)
    dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:162)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
    dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
    scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
    dotty.tools.dotc.Run.runPhases$5(Run.scala:172)
    dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:180)
    dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
    dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:65)
    dotty.tools.dotc.Run.compileUnits(Run.scala:187)
    dotty.tools.dotc.Run.compileSources(Run.scala:124)
    dotty.tools.dotc.Run.compile(Run.scala:107)
    dotty.tools.dotc.Driver.doCompile(Driver.scala:36)
    dotty.tools.dotc.Driver.process(Driver.scala:189)
    dotty.tools.dotc.Main.process(Main.scala)
    xsbt.CachedCompilerImpl.run(CachedCompilerImpl.java:69)
    xsbt.CompilerInterface.run(CompilerInterface.java:41)
    java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```
</details>

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -683,15 +683,28 @@ object desugar {
           val mods = constr1.mods
           mods.is(Private) || (!mods.is(Protected) && mods.hasPrivateWithin)
         }
+ 
+        /** Does one of the parameter's types (in the first param clause)
+         *  mention a preceding parameter?
+         */
+        def isParamDependent = constrVparamss match
+          case vparams :: _ =>
+            val paramNames = vparams.map(_.name).toSet
+            vparams.exists(_.tpt.existsSubTree {
+              case Ident(name: TermName) => paramNames.contains(name)
+              case _ => false
+            })
+          case _ => false
 
         val companionParent =
-          if (constrTparams.nonEmpty ||
-              constrVparamss.length > 1 ||
-              mods.is(Abstract) ||
-              restrictedAccess ||
-              isEnumCase) anyRef
+          if constrTparams.nonEmpty
+             || constrVparamss.length > 1
+             || mods.is(Abstract)
+             || restrictedAccess
+             || isParamDependent
+             || isEnumCase
+          then anyRef
           else
-            // todo: also use anyRef if constructor has a dependent method type (or rule that out)!
             constrVparamss.foldRight(classTypeRef)((vparams, restpe) => Function(vparams map (_.tpt), restpe))
         def widenedCreatorExpr =
           widenDefs.foldLeft(creatorExpr)((rhs, meth) => Apply(Ident(meth.name), rhs :: Nil))

--- a/compiler/src/dotty/tools/dotc/ast/untpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/untpd.scala
@@ -747,4 +747,12 @@ object untpd extends Trees.Instance[Untyped] with UntypedTreeInfo {
   class UntypedDeepFolder[X](f: (X, Tree) => X) extends UntypedTreeAccumulator[X] {
     def apply(x: X, tree: Tree)(implicit ctx: Context): X = foldOver(f(x, tree), tree)
   }
+
+  /** Is there a subtree of this tree that satisfies predicate `p`? */
+  def (tree: Tree).existsSubTree(p: Tree => Boolean)(implicit ctx: Context): Boolean = {
+    val acc = new UntypedTreeAccumulator[Boolean] {
+      def apply(x: Boolean, t: Tree)(implicit ctx: Context) = x || p(t) || foldOver(x, t)
+    }
+    acc(false, tree)
+  }
 }

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -1130,8 +1130,13 @@ class Namer { typer: Typer =>
 
       index(constr)
       index(rest)(localCtx)
-      symbolOfTree(constr).ensureCompleted()
-
+      symbolOfTree(constr).info.stripPoly match // Completes constr symbol as a side effect
+        case mt: MethodType if cls.is(Case) && mt.isParamDependent =>
+          // See issue #8073 for background
+          ctx.error(
+              i"""Implementation restriction: case classes cannot have dependencies between parameters""",
+              cls.sourcePos)
+        case _ =>
       denot.info = savedInfo
     }
 

--- a/tests/neg/i8069.scala
+++ b/tests/neg/i8069.scala
@@ -1,0 +1,8 @@
+trait A
+  type B
+
+enum Test
+  case Test(a: A, b: a.B)   // error: Implementation restriction: case classes cannot have dependencies between parameters
+
+case class Test2(a: A, b: a.B) // error: Implementation restriction: case classes cannot have dependencies between parameters
+

--- a/tests/pos/i8069.scala
+++ b/tests/pos/i8069.scala
@@ -1,0 +1,12 @@
+class A {
+  class B
+}
+
+class C(val a: A, val b: a.B)
+
+@main def Test =
+  val a = A()
+  val b = a.B()
+  val c = C(a, b)
+  val d = c.b
+  val d1: c.a.B = d

